File: apps/web/src/features/expenses/api/list-expenses.ts
import { api } from '@/api/client'
import type { ExpensePage } from '@/features/expenses/expense'

export async function listExpenses(
  month: string,
  limit: number,
  offset: number,
): Promise<ExpensePage> {
  const { data, error } = await api.GET('/api/v1/expenses', {
    params: { query: { month, limit, offset } },
  })
  if (error) throw new Error('Failed to load expenses')
  return data
}

File: apps/api/internal/expenses/application/list_expenses.go
package application

import (
	"context"

	"caudal-api/internal/expenses/ports"
	shared "caudal-api/internal/shared/domain"
)

type ListExpenses struct {
	repository ports.Repository
}

func NewListExpenses(repository ports.Repository) ListExpenses {
	return ListExpenses{repository: repository}
}

func (useCase ListExpenses) Execute(ctx context.Context, query ListExpensesQuery) (ports.Page, error) {
	start, end := shared.FinancialMonthBounds(query.Year, query.Month)
	return useCase.repository.List(ctx, start, end, query.Limit, query.Offset)
}
